src/ILCompiler.Compiler/src/Compiler/NameMangler.cs

```diff
@@ -74,7 +74,15 @@ internal string SanitizeName(string s, bool typeName = false)
                 // TODO: We assume that there won't be collisions with our own or C++ built-in identifiers.
                 sb.Append("_");
             }
-            return (sb != null) ? sb.ToString() : s;
+
+            string santizedName = (sb != null) ? sb.ToString() : s;
+
+            // The character sequences denoting generic instantiations, arrays, byrefs, or pointers must be
+            // restricted to that use only. Replace them if they happened to be used in any identifiers in 
+            // the compilation input.
+            return _mangleForCplusPlus
+                ? santizedName.Replace(EnterNameScopeSequence, "_AA_").Replace(ExitNameScopeSequence, "_VV_")
```

Comment:
I was trying to keep the C++ to ASCII as much as possible, partially because different compilers have different levels of support for unicode character identifiers and this emitted C++ code is supposed to be ultra portable. I'll do a quick survey of the current support in gcc / llvm without requiring `-fextended-identifiers` switch and see what we can expect to work broadly.